### PR TITLE
Backup code authenticator initial PR

### PR DESCRIPTION
**Pull request by bhagyasakalanka**

### CODE REVIEW

File: component/authenticator/src/main/java/org/wso2/carbon/identity/application/authenticator/backupcode/BackupCodeAuthenticator.java [hunk 1]
```diff
@@ -0,0 +1,881 @@
...
+
+        String username = null;
+        String tenantDomain = context.getTenantDomain();
+        context.setProperty(AUTHENTICATION,
+                BackupCodeAuthenticatorConstants.BACKUP_CODE_AUTHENTICATOR_NAME);
+        if (!tenantDomain.equals(BackupCodeAuthenticatorConstants.SUPER_TENANT_DOMAIN)) {
+            IdentityHelperUtil
+                    .loadApplicationAuthenticationXMLFromRegistry(context, getName(), tenantDomain);
+        }
+        AuthenticatedUser authenticatedUserFromContext = BackupCodeUtil.getAuthenticatedUser(context);
```
Comment (JKAUSHALYA) [inline component/authenticator/src/main/java/org/wso2/carbon/identity/application/authenticator/backupcode/BackupCodeAuthenticator.java:161]: Find whether there is already an utility method to do this.

File: component/authenticator/src/main/java/org/wso2/carbon/identity/application/authenticator/backupcode/BackupCodeAuthenticator.java [hunk 2]
```diff
@@ -0,0 +1,881 @@
...
+        if (authenticatedUserFromContext == null) {
+            throw new AuthenticationFailedException(
+                    BackupCodeAuthenticatorConstants.ErrorMessages.ERROR_CODE_NO_AUTHENTICATED_USER.getCode(),
+                    BackupCodeAuthenticatorConstants.ErrorMessages.ERROR_CODE_NO_AUTHENTICATED_USER.getMessage());
+        }
+
+        /*
+        The username that the server is using to identify the user, is needed to be identified, as
+        for the federated users, the username in the authentication context may not be same as the
+        username when the user is provisioned to the server.
```
Comment (JKAUSHALYA) [inline component/authenticator/src/main/java/org/wso2/carbon/identity/application/authenticator/backupcode/BackupCodeAuthenticator.java]: Use following format.
/*
\* Your comment. 
\*
*/

Comment (JKAUSHALYA) [reply component/authenticator/src/main/java/org/wso2/carbon/identity/application/authenticator/backupcode/BackupCodeAuthenticator.java]: Update every instance.

File: component/authenticator/src/main/java/org/wso2/carbon/identity/application/authenticator/backupcode/BackupCodeAuthenticator.java [hunk 3]
```diff
@@ -0,0 +1,881 @@
...
+            if (log.isDebugEnabled()) {
+                log.debug("Backup Code  is enabled by admin: " + isBackupCodesEnabledByAdmin);
+            }
+            // This multi option URI is used to navigate back to multi option page to select a different
+            // authentication option from backup code pages.
+            String multiOptionURI = BackupCodeUtil.getMultiOptionURIQueryParam(request);
+
+            if (isBackupCodesExistForUser &&
+                    request.getParameter(BackupCodeAuthenticatorConstants.ENABLE_BACKUP_CODE) == null) {
+                //if backup code is enabled for the user.
```
Comment (JKAUSHALYA) [inline component/authenticator/src/main/java/org/wso2/carbon/identity/application/authenticator/backupcode/BackupCodeAuthenticator.java]: Use correct format.

File: component/authenticator/src/main/java/org/wso2/carbon/identity/application/authenticator/backupcode/BackupCodeAuthenticator.java [hunk 4]
```diff
@@ -0,0 +1,881 @@
...
+                String backupCodeLoginPageUrl =
+                        buildBackupCodeLoginPageURL(context, username, retryParam, multiOptionURI);
+                response.sendRedirect(backupCodeLoginPageUrl);
+            } else {
+
+                if (isBackupCodesEnabledByAdmin) {
```
Comment (JKAUSHALYA) [inline component/authenticator/src/main/java/org/wso2/carbon/identity/application/authenticator/backupcode/BackupCodeAuthenticator.java]: Remove this condition as we have no such thing.

File: component/authenticator/src/main/java/org/wso2/carbon/identity/application/authenticator/backupcode/BackupCodeAuthenticator.java (hunk 3 above)
Comment (JKAUSHALYA) [reply component/authenticator/src/main/java/org/wso2/carbon/identity/application/authenticator/backupcode/BackupCodeAuthenticator.java]: Update all the instances.

File: component/authenticator/src/main/java/org/wso2/carbon/identity/application/authenticator/backupcode/BackupCodeAuthenticator.java [hunk 5]
```diff
@@ -0,0 +1,881 @@
...
+                authenticatedUser.setAuthenticatedSubjectIdentifier(username);
+                authenticatedUser.setUserName(UserCoreUtil.removeDomainFromName(
+                        MultitenantUtils.getTenantAwareUsername(username)));
+                authenticatedUser.setUserStoreDomain(UserCoreUtil.extractDomainFromName(username));
+                authenticatedUser.setTenantDomain(MultitenantUtils.getTenantDomain(username));
+                context.setSubject(authenticatedUser);
+            } else {
+                context.setSubject(AuthenticatedUser.createLocalAuthenticatedUserFromSubjectIdentifier(username));
+            }
+        } catch (NumberFormatException e) {
```
Comment (JKAUSHALYA) [inline component/authenticator/src/main/java/org/wso2/carbon/identity/application/authenticator/backupcode/BackupCodeAuthenticator.java]: Do we need this? Do we need to treat this as a number? Just comparing as string is enough AFAIT.

File: component/authenticator/src/main/java/org/wso2/carbon/identity/application/authenticator/backupcode/BackupCodeAuthenticator.java (hunk 1 above)
Comment (bhagyasakalanka) [reply component/authenticator/src/main/java/org/wso2/carbon/identity/application/authenticator/backupcode/BackupCodeAuthenticator.java:161]: There are no public methods for this

File: component/authenticator/src/main/java/org/wso2/carbon/identity/application/authenticator/backupcode/BackupCodeAuthenticator.java (hunk 2 above)
Comment (bhagyasakalanka) [reply component/authenticator/src/main/java/org/wso2/carbon/identity/application/authenticator/backupcode/BackupCodeAuthenticator.java]: Fixed

File: component/authenticator/src/main/java/org/wso2/carbon/identity/application/authenticator/backupcode/BackupCodeAuthenticator.java (hunk 4 above)
Comment (bhagyasakalanka) [reply component/authenticator/src/main/java/org/wso2/carbon/identity/application/authenticator/backupcode/BackupCodeAuthenticator.java]: Fixed

File: component/authenticator/src/main/java/org/wso2/carbon/identity/application/authenticator/backupcode/BackupCodeAuthenticator.java (hunk 5 above)
Comment (bhagyasakalanka) [reply component/authenticator/src/main/java/org/wso2/carbon/identity/application/authenticator/backupcode/BackupCodeAuthenticator.java]: Yea fixed that

File: component/authenticator/src/main/java/org/wso2/carbon/identity/application/authenticator/backupcode/BackupCodeAuthenticator.java (hunk 3 above)
Comment (bhagyasakalanka) [reply component/authenticator/src/main/java/org/wso2/carbon/identity/application/authenticator/backupcode/BackupCodeAuthenticator.java]: Fixed

**Review by JKAUSHALYA — APPROVED** [review]

**Pull request merged**